### PR TITLE
gh-94017: Improve clarity of sqlite3 transaction handling docs

### LINKED ISSUE #94017: Improve clarity of sqlite3 transaction handling docs
**Documentation**

The sqlite3 module docs has a section called [Controlling Transactions](https://docs.python.org/3/library/sqlite3.html#controlling-transactions).

There are some improvement possibilities:

- The current docs start out by describing the _SQLite_ autocommit mode. This _can_ be confusing to new users, especially since the DB API (PEP-249) describes its own autocommit behaviour. Suggesting to tone down the SQLite autocommit mode and instead focus on what the sqlite3 module does.
- The docs say that implicit BEGIN statements are issues before DML statements. This is partly true. Implicit BEGIN statements are issued before INSERT, UPDATE, DELETE, and REPLACE statements _when_ executed by `execute` or `executemany`, but not by `executescript`. Other DML statements do not receive such treatment.
- We should update the docs of the `execute`, `executemany`, and `executescript` cursor methods to briefly mention their transaction control peculiarities.
- Since _SQLite autocommit mode_ is mentioned, we should note that it is in fact possible to query this (low-level) mode using the `in_transaction` connection attribute. Such information can help developers debug transaction control problems.
- Possibly augment the docs of the `isolation_level` attribute. Currently, the `isolation_level` _property_ of `connection` refer to the docs of the `isolation_level` _attribute_, which again refer to the _Controlling Transactions_ section. The `isolation_level` property should have a concise description (reference), and link to the `isolation_level` attribute. The `isolation_level` attribute should provide a _reference_ for what it does, and briefly mention expected behaviour. The _Controlling Transactions_ section should provide more background (explanation) to what is happening, and it should briefly mention PEP 249, and the SQLite autocommit mode.
- Revisit any transaction control examples and possibly turn them into clearer how-to's, or guides.

### PR DESCRIPTION
**Pull request by erlend-aasland**
Fixes gh-94017

### CODE REVIEW

**Comment by erlend-aasland** [comment]
cc. @CAM-Gerlach, @AlexWaygood. Would you two mind reviewing this? I'll go over it later today, to be sure the docs align with the actual implementation. Other that, I worry that the new tone of the docs may come across as ... perhaps a trifle strict, or negative sounding (in lack of more appropriate words). I do hope this makes things clearer, though.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: If we say A, we must also say B? It feels like we give the users two loose ends here: there's a PEP 249 autocommit concept, and there's an SQLite autocommit concept; now, go figure it out yourselves :) Is that acceptable? I neither want to repeat PEP 249 stuff here, nor the SQLite docs. A possible enhancement is to add links to the relevant sections in the respcective docs.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: Feels weird to start a sentence with "".

```suggestion
      Both "" and "DEFERRED" carry the same meaning;
      they imply deferred isolation level.
```

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: "current default" sounds somewhat confusing to me, as it seems a little oxymoronic. is there maybe another way to describe this that doesn't involve putting both those words together?

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
      or to one of the ``"DEFERRED"``, ``"IMMEDIATE"`` or ``"EXCLUSIVE"``
      transaction-handling modes.
```

Parallel the first clause of the sentence in stating the meaning of these values, and save the mention of the alias for `DEFERRED` until when the default is introduced, to avoid confusion. Also, use correct reST syntax.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
      If not overridden by the *isolation_level* parameter of :func:`connect`,
      the default is ``""``, which is an alias for ``"DEFERRED"``.
```

Say the same thing but more simply, directly and succinctly, and match the above. Also, fix reST syntax.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: Here, we should either link to the section below that discusses it (if it does), or else directly to SQLite's own autocommit docs.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
      If :attr:`Connection.isolation_level` is not :const:`None`,
```

This is an attribute of `Connection`, not `Cursor`, and so the link doesn't resolve.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
by :pep:`249`.
```

Link it.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
If the connection attribute :attr:`~Connection.isolation_level`
is not :const:`None`,
```

Again, `Connection` needs to be specified as the class, or the link won't work.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
new transactions are implicitly opened before
```

The above phrase adds verbosity, complexity and confusion from the multiple levels of colons without really adding anything to meaning

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
:meth:`~Cursor.execute` or :meth:`~Cursor.executemany` executes
INSERT, UPDATE, DELETE or REPLACE statements.
```

IMO, the relatively small benefit of making these relatively few and short terms a bulleted list does not outweigh the complexity, verbosity and indirectness cost of doing so. Therefore, I suggest a simpler and more direct prose approach.

Also, fix conjunction/plural agreement.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
The :meth:`~Cursor.executescript` method implicitly commits
any pending transaction before execution of the given SQL script.
```

Avoid potentially confusing "additionally", since the statements here are parallel to, not additive to the statements about the other two methods in the previous paragraph. Also, state the method it applies to out front so the distinction between the two paras is clearer, and this construction is a bit shorter, simpler, more direct and easier to understand, and avoids an extra comma.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
Use the :meth:`~Connection.commit` and :meth:`~Connection.rollback` methods
```

Plural agreement

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
executes via the :attr:`Connection.isolation_level` attribute.
```

Fix broken link

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
setting :attr:`Connection.isolation_level` to :const:`None`.
```

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
but also allows the user to perform their own transaction handling
using explicit SQL statements.
```

Clearer and more appropriate verbiage

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
:attr:`Connection.in_transaction` attribute.
```

Fix link

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
   :pep:`249`'s autocommit concept must not be mistaken for
   SQLite's autocommit mode.
```

Add PEP link

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: I added an explicit link to PEP 249; we could link the specific section using the syntax ``:pep:`PEP 249 <249#section-id>` ``.

IMO yes, we should also link the appropriate section in the SQLite docs here, rather than leaving the reader completely out to dry.

**Review by CAM-Gerlach — COMMENTED** [review]
Thanks @erlend-aasland , this is a definite improvement. I do have some clarity, textual and reST fixes and suggestions.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Hah, I didn't notice that one until you pointed it out 😆 I think it survived from the original version.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Ah, I was under the impression that it would magically find its way :) I'll be more explicit.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Indeed it did :) —but I figured I'd mention it while you were at it.

File: Doc/library/sqlite3.rst
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]:
```suggestion
      *sql* is an ``INSERT``, ``UPDATE``, ``DELETE``, or ``REPLACE`` statement,
```

File: Doc/library/sqlite3.rst
Comment (AlexWaygood) [reply Doc/library/sqlite3.rst]: I'd also put all these (INSERT, UPDATE, DELETE, REPLACE) in double-backticks, personally.

**Review by AlexWaygood — COMMENTED** [review]
+1 to most of Cam's thoughts

**Comment by erlend-aasland** [comment]
Thanks, both of you; highly appreciated. I'm finally getting around to address the reviews.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Actually, I think we should defer adding this note until #93823 has landed. I'll remove it for now.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I just removed "current default". However, we might as well just remove the whole sentence. It does not add much value.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: This is a good catch. The "in addition" would definitely add confusion.

**Comment by erlend-aasland** [comment]
Reviews addressed in a530b3e8034e698575faea8628fa9a51b1ca5cca.
I mostly went for `:attr:\`~Connection.isolation_level\`` style links instead of the more explicit `:attr:\`Connection.isolation_level\`` style links. I can use the latter if that's preferred.

**Comment by CAM-Gerlach** [comment]
> I mostly went for ~Connection.isolation_level style links instead of the more explicit Connection.isolation_level style links. I can use the latter if that's preferred.

That's fine with me, since it isn't used on multiple classes in the module and links directly to them in either case. I wasn't very consistent about using that myself.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst:881]: To note, this never seems to explicitly describe what `parameters` are or in what form they should be passed, though perhaps out of scope here.

**Review by CAM-Gerlach — APPROVED** [review]
LGTM, thanks @erlend-aasland 

I had one comment, but that might be out of scope here and better addressed in a future PR.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst:881]: Lets do that in another PR.

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I'll remove this first sentence. But I want to do some final adjustments on the `isolation_level` docs before landing this PR. Bear with me :)

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [inline Doc/library/sqlite3.rst]: There is one thing that still disturbs me with this part: it is not technically correct to say that DEFERRED, IMMEDIATE, and EXCLUSIVE are transaction handling modes; this may be percieved as if sqlite3 has four transaction handling modes (one for each allowed value). However, it has only two:

1. *isolation_level* is `None`: no implicit transaction handling is performed (except for the implicit commit in `executescript`)
2. *isolation_level* is a string: implicit transaction handling is performed (implicit BEGIN before INSERT, DELETE, etc., and of course the commit in executescript.)

Perhaps it is better to describe these as:

1. *isolation_level* is `None`: transactions are never implicitly opened
2. *isolation_level* is `str`: transactions are implicitly opened before INSERT, etc.

The three allowed string values relates to low level SQLite [transaction behaviour](https://www.sqlite.org/lang_transaction.html), but the *sqlite3* implicit transaction mode is the same, no matter the *SQLite* transaction *behaviour*.

I'm not sure how to best communicate these things, but we cannot land this PR yet :(

Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: I tried to address all of this in 1473e948d54b18755dcef8daf421e11f48d7ebf7.

**Comment by erlend-aasland** [comment]
I'm sorry, but I did a last rewrite of things. Hopefully to the better. I'd be happy if you could review commit 1473e948d54b18755dcef8daf421e11f48d7ebf7. If it was for the worse, I'll revert it.

For the "Controlling Transactions" section, I tried to improve the visual separation of the two *sqlite3* transaction modes.

File: Doc/library/sqlite3.rst
Comment (AlexWaygood) [inline Doc/library/sqlite3.rst]:
```suggestion
You can choose the underlying SQLite `transaction behaviour`_,
```

**Comment by CAM-Gerlach** [comment]
> I'm sorry, but I did a last rewrite of things, trying to address [#94320 (comment)](https://github.com/python/cpython/pull/94320#discussion_r912403557). Hopefully to the better. I'd be happy if you could review commit [1473e94](https://github.com/python/cpython/commit/1473e948d54b18755dcef8daf421e11f48d7ebf7). If it was for the worse, I'll revert it.

No need to apologize for doing more work and making the docs better! I'll review it now—sorry myself for the delay; it was a holiday weekend and I was with family.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
      corresponding to the underlying `SQLite transaction behaviour`_,
```

Clearer/cleaner phrasing and include SQLite in the link target, since it is a key part of the target's name

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]: It seems this is the only use of the term "DML statement" in this document, and neither is it defined in the linked SQLite docs. I'm no database expert, though I have read most of these docs, but still had to go Google it myself and read through the linked resources to determine what statements were affected (with some ambiguity about `SELECT`).

Unless nearly all readers would be expected to know what it means and what statements (presumably `SELECT`, `INSERT`, `UPDATE` and `DELETE`) are included, I suggest explicitly specifying _which_ statements, and/or linking to a definition/explanation of the term.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
.. _SQLite transaction behaviour: https://www.sqlite.org/lang_transaction.html#deferred_immediate_and_exclusive_transactions
```

As above

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
You can choose the underlying `SQLite transaction behaviour`_ —
```

As above, "SQLite" is an inherent part of the target name here. Also, avoid a comma splice and set this off with an em dash parenthetical.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
that is, whether and what type of ``BEGIN`` statements
the :mod:`sqlite3` module implicitly executes —
```

Be more clear that it not only controls the type of `BEGIN` statement, but whether they are executed at all or the underlying `autocommit` mode is relied upon, as well as that we are referring to the `sqlite3` _Python module_ rather than the underlying SQLite database backend.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [inline Doc/library/sqlite3.rst]:
```suggestion
regardless of the value of :attr:`~Connection.isolation_level`.
```

Editing typo

**Review by CAM-Gerlach — COMMENTED** [review]
Review feedback on the final revisisions

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: +1

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: Hm, how about just simplifying it and linking to the explanation?

```suggestion
      some implicit :ref:`transaction management
      <sqlite3-controlling-transactions>` is performed.
```

File: Doc/library/sqlite3.rst
Comment (erlend-aasland) [reply Doc/library/sqlite3.rst]: +1

But I'll remove the unneeded "self reference"; no need to link to the sqlite3 docs when you're already there.

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Looks good overall, I just suggest eliding "some" since it is rather vaugue and confusing, and sounds strange in that context in English technical writing:

```suggestion
      implicit :ref:`transaction management
      <sqlite3-controlling-transactions>` is performed.
```

File: Doc/library/sqlite3.rst
Comment (CAM-Gerlach) [reply Doc/library/sqlite3.rst]: Its more for conceptual disambiguity of what's being referring to and formatting consistency of the rendered output than the need for the reference, but its not a big deal.

**Comment by erlend-aasland** [comment]
Thanks again for the reviews; I'll land this later tonight.

**Pull request merged**
